### DISCUSSION TITLE
Pydantic validator executed twice if more than one body is allowed

### DISCUSSION THREAD

**Discussion by JavierSanchezCastro** [fastapi/fastapi#9535] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, status
from pydantic import BaseModel, condecimal, Field, validator
from enum import Enum


class Expense_Type(Enum):
    EQUAL_SHARES = "Equal Shares"
    PERCENTAGE = "Percentage"

class ExpenseBase(BaseModel):
    name: str = Field(max_length=25)
    description: str | None = Field(max_length=512)
    amount: condecimal(max_digits=10, decimal_places=2, gt=0)

class CreateExpense(ExpenseBase):
    pass

class CreateExpenseEqual(CreateExpense):
    users: list[str]

    @validator('users', pre=True)
    def validate_unique_users(cls, users):
        if len(set(users)) != len(users):
            raise ValueError("Repeated users found in list.")
        return users

class CreateExpensePercentage(CreateExpense):
    users: dict[str, float]
    other_field: int

app = FastAPI()

@app.post("/{group_id}/expense", status_code=status.HTTP_201_CREATED)
def create_expense(group_id: int, type: Expense_Type, expense: CreateExpenseEqual | CreateExpensePercentage):
    return None
```

### Description

If I send this as body I don't receive any error

```
{
  "name": "string",
  "description": "string",
  "amount": 1,
  "users": ["string"]
}
```

But If I send this, I'm gonna see multiple errors
```
{
  "name": "string",
  "description": "string",
  "amount": 1,
  "users": ["string", "string"]
}
```
```
{
  "detail": [
    {
      "loc": [
        "body",
        "users"
      ],
      "msg": "Repeated users found in list.",
      "type": "value_error"
    },
    {
      "loc": [
        "body",
        "users"
      ],
      "msg": "value is not a valid dict",
      "type": "type_error.dict"
    },
    {
      "loc": [
        "body",
        "other_field"
      ],
      "msg": "field required",
      "type": "value_error.missing"
    }
  ]
}
```
One error for the custom validator, another one for type_error dict from CreateExpensePercentage, and I add "other_field" to be sure that these two errors are coming from this BaseModel class.

I see some discussions such as #8659 talking about something related.

Is there any way to get only the error from the body I send?
Am I doing well having one endpoint with multiple body allowed? Is there a better way to do/improve it?

### Operating System

Linux

### Operating System Details

Ubuntu

### FastAPI Version

0.95.1

### Python Version

Python 3.10.6

**Comment by yinziyan1206** [comment]
I think it is hard for `fastapi` to find out which type your body should be.

**Comment by YuriiMotov** [comment, marked as answer]
That's because when you use `Union`, FastAPI tries to validate input with all types within union. If it fails for first it will go to next.
You can avoid this using discriminated unions
